src/apps/index.ts

```diff
@@ -37,7 +37,7 @@ export async function index(): Promise<Array<App>> {
   })
 
   const data = await response.json();
-  return data
+  return Object.values(data).map(d => ({ domain: (d as Array<string>)[0] }))
```

Comment:
This won't work for what I get returned from the API:
```json
{
    "427": [
        "elegant-small-old-wolf.fission.app"
    ],
    "1291": [
        "victorious-short-azure-witch.fission.app"
    ],
    "1292": [
        "thin-azure-ghost.fission.app"
    ],
    "1293": [
        "petite-oval-teal-tulip.fission.app"
    ],
    "1294": [
        "web-uploaded-app.fission.app"
    ],
    "1295": []
}
```
The "1295" key contains an empty list.

Personally, I'd rather change the type than the implementation. I feel like there's a reason why the backend returns a dictionary of lists, and maybe we should reflect that in the API of webnative.

Thoughts?

Reply:
Oh ok, thanks, I missed that. What's the reasoning here with the keys and lists? (ie. what is this key exactly and why are we working with lists)

Reply:
It was on my plate to figure this out for a long time now, but I haven't gotten to it, yet. So: I don't know.

I just talked to James. It might make sense to just assume that there's one domain per app (not zero and not more than that), and the empty list *might* be a bug.

We *could* design the API to be more future-proof, though: At some point, there's going to be multiple domains per app.

However, e.g. the dashboard is making the assumption that there's one domain per app at the moment (and filtering out apps without domains), so :man_shrugging: 

I'll take a look at the server.

Reply:
There needs to be a "main" domain per app, that always works. That is the `*.fission.app` subdomain. So keep it with that for now, and when we have automated addition of custom domains.

Reply:
Ok, so I took a look.
Here's the code for the `.index()` handler:</summary>
```hs
index :: (MonadDB t m, App.Retriever t, App.Domain.Retriever t) => ServerT API.App.Index m
index Authorization {about = Entity userId _} = runDB do
  apps        <- App.ownedBy userId
  appXDomains <- forM apps findDomains
  let normalized = (\(k, v) -> (fromIntegral (fromSqlKey k), v)) <$> appXDomains
  return (Map.fromList normalized)
  where
    findDomains :: App.Domain.Retriever m => Entity App -> m (AppId, [URL])
    findDomains (Entity appId _) = do
      appDomains <- App.Domain.allForApp appId
      return (appId, toURL <$> appDomains)

    toURL :: Entity AppDomain -> URL
    toURL (Entity _ AppDomain {..}) = URL appDomainDomainName appDomainSubdomain
```

So it's a dictionary of apps by SQL key with a list of associated domains for each.

Also, there's two things you can destroy via the REST API: Apps by domain and apps by ID. If you destroy by domain, you basically just remove the domain entry from the app's list of domains. However, **it's still there**. If you want to, you can re-add a domain to it via the ID you get from `.index()`.
If you destroy it by ID, it's just gone, but that's not an API that webnative exposes at the moment.

More related code in `fission-web-server/library/Fission/Web/Server/Handler/App/Destroy.hs` and `fission-web-server/library/Fission/Web/Server/App/Destroyer/Class.hs`
```hs
handler :: (MonadTime m, MonadThrow m, MonadLogger m, App.Destroyer m) => ServerT API.App.Destroy m
handler = destroyByURL :<|> destroyById

destroyByURL :: (MonadTime m, MonadThrow m, MonadLogger m, App.Destroyer m) => ServerT API.App.Destroy.ByURL m
destroyByURL URL {..} Authorization {about = Entity userId _} = do
  now <- currentTime
  Web.Error.ensureM $ App.destroyByURL userId domainName subdomain now
  return NoContent

destroyById :: (MonadTime m, MonadThrow m, MonadLogger m, App.Destroyer m) => ServerT API.App.Destroy.ById m
destroyById appId Authorization {about = Entity userId _} = do
  now <- currentTime
  Web.Error.ensureM $ App.destroy userId (toSqlKey $ fromIntegral appId) now
  return NoContent
```
and
```hs
class Monad m => Destroyer m where
  destroy ::
       UserId
    -> AppId
    -> UTCTime
    -> m (Either Errors' [URL])

  destroyByURL ::
       UserId
    -> DomainName
    -> Maybe Subdomain
    -> UTCTime
    -> m (Either Errors' [URL])

instance MonadIO m => Destroyer (Transaction m) where
  destroy userId appId now = do
    appDomains <- AppDomain.allForApp appId
    destroyAssociated userId appId appDomains now

  destroyByURL userId domainName maySubdomain now =
    AppDomain.allSiblingsByDomain domainName maySubdomain >>= \case
      Left  errs ->
        return $ relaxedLeft errs

      Right [] ->
        return . openLeft $ NotFound @AppDomain

      Right appDomains@(Entity _ AppDomain {appDomainAppId} : _) ->
        destroyAssociated userId appDomainAppId appDomains now

destroyAssociated ::
  MonadIO m
  => UserId
  -> AppId
  -> [Entity AppDomain]
  -> UTCTime
  -> Transaction m (Either Errors' [URL])
destroyAssociated userId appId appDomains now =
  App.byId userId appId >>= \case
    Left err ->
      return $ relaxedLeft err

    Right (Entity _ app) ->
      if isOwnedBy userId app
        then do
          putMany (toEvent now <$> appDomains)
          deleteCascade appId
          return $ Right (extractURL <$> appDomains)
        else
          return . openLeft $ ActionNotAuthorized @App userId
```

---

I think it makes sense to reflect this API soon? But probably in another PR. I was planning on overhauling the API from what I've learned working with it in the dashboard anyway.

Reply:
Ah ok, sounds good 👍  I can change it in this PR, so basically instead of the `{ domain }` response, you'd like `{ appId, domains }`?

Reply:
> I just talked to James. It might make sense to just assume that there's one domain per app (not zero and not more than that), and the empty list _might_ be a bug.

Yeah, as @bmann mentioned, we should always have one stable domain (APP_NAME.fission.app) but we'll have additional custom domains eventually too.

Reply:
> Ah ok, sounds good +1 I can change it in this PR, so basically instead of the { domain } response, you'd like { appId, domains }?

I think it's likely that I'd touch the API with a PR of my own as a result of finishing the current dashboard slice :sweat_smile: 

> Yeah, as @bmann mentioned, we should always have one stable domain (APP_NAME.fission.app) but we'll have additional custom domains eventually too.

Hm. I agree that this would be nice, but at the moment it's pretty easy to break this invariant using the REST API. What do we do? Can we assume this invariant to hold if it's breakable, or do we roll with the way it works right now? Or do we fix the server, too? Or do we enforce the invariant, but only in webnative?

Sorry, that's starting to get quite off-topic. I think this is not a problem with the this PR anymore. :smile:

Reply:
> I think it's likely that I'd touch the API with a PR of my own

Sure thing, I'll just do a quick fix then, filtering out the empty lists.